python_modules/dagster/dagster/_core/execution/context/system.py

```diff
@@ -975,26 +977,21 @@ def partition_mapping_for_input(self, input_name: str) -> Optional[PartitionMapp
         else:
             return None
 
-    def _get_input_asset_event(self, key: AssetKey, retries: int = 0) -> Optional["EventLogRecord"]:
-        from dagster._core.definitions.data_version import (
-            extract_data_version_from_entry,
-        )
-
-        is_step_internal_input = key in self._data_version_cache
+    def _get_input_asset_event(self, key: AssetKey) -> Optional["EventLogRecord"]:
         event = self.instance.get_latest_data_version_record(key)
-        if event is None and is_step_internal_input and retries <= 5:
-            return self._get_input_asset_event(key, retries + 1)
-        elif event is None:
-            return None
-        else:
-            # The cache gets populated for inputs generated during the current step. We need to keep
-            # reloading the latest event until the data version matches the one in the cache to ensure
-            # the storage id reflects the input actually used.
-            data_version = extract_data_version_from_entry(event.event_log_entry)
-            if is_step_internal_input and data_version != self._data_version_cache[key]:
-                return self._get_input_asset_event(key, retries + 1)
-            else:
-                return event
+        if event:
+            self._check_input_asset_event(key, event)
+        return event
+
+    def _check_input_asset_event(self, key: AssetKey, event: EventLogRecord) -> None:
+        assert event.event_log_entry
+        event_data_version = extract_data_version_from_entry(event.event_log_entry)
+        if key in self._data_version_cache and self._data_version_cache[key] != event_data_version:
+            warnings.warn(
+                f"Data version mismatch for asset {key}. Data version from materialization"
+                f" within current step is `{self._data_version_cache[key]}`. Most recent"
+                " materialization will be used for provenance tracking."
+            )
```

Comment:
nit: should we print out the most recent data version as well?

Reply:
oops thought I was doing this already